src/go/k8s/pkg/networking/address_types.go

```diff
@@ -0,0 +1,57 @@
+package networking
+
+import (
+	"log"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+const defaultType = corev1.NodeExternalIP
+
+var typePriority = []corev1.NodeAddressType{
+	corev1.NodeExternalDNS,
+	corev1.NodeExternalIP,
+	corev1.NodeInternalDNS,
+	corev1.NodeInternalIP,
+	corev1.NodeHostName,
+}
+
+// GetPreferredAddress returns the preferred node address if
+// it is available and a preferred type is provided. Otherwise,
+// it loops over the address types by priority and returns the
+// first one that is available.
+func GetPreferredAddress(
+	node *corev1.Node, preferredType corev1.NodeAddressType,
+) string {
+	if node == nil {
+		return ""
+	}
+	if preferredType == "" {
+		preferredType = defaultType
+	}
+	// Return the address of the preferred type, if found
+	if s := findAddressType(node, preferredType); s != "" {
+		log.Println("address:", s, "type:", preferredType)
+		return s
+	}
+	// When the preferred address type is unavailable, go through
+	// the types of addresses following the predefined type priority.
+	for _, t := range typePriority {
```

Comment:
this is kind of confusing. So in config I say "I want this preferred type" but then you assign other if it's not available? 🤔

Reply:
I decided to change the approach (updated the description to reflect it). I'll keep the logic simpler by not falling back to an alternative address type and see if there are enough use cases for the "priority" approach. It's fairly easy to add.

For the sake of discussion only, it can be "confusing" as you said. However, I see potential advantages like robustness of the spec across environments (e.g., backup yamls and restore elsewhere) and ease of deployment in a quickstart scenario where the user doesn't know what address types are available.